WalletWasabi/Services/WalletService.cs

```diff
@@ -315,29 +322,27 @@ public HdPubKey GetReceiveKey(string label, IEnumerable<HdPubKey> dontTouch = nu
 			return ret;
 		}
 
-		public List<SmartCoin> GetHistory(SmartCoin coin, IEnumerable<SmartCoin> current)
+		public List<SmartCoin> GetHistory(SmartCoin coin, List<SmartCoin> current, ILookup<Script, SmartCoin> lookupScriptPubKey, ILookup<uint256, SmartCoin> lookupSpenderTransactionId, ILookup<uint256, SmartCoin> lookupTransactionId)
 		{
 			Guard.NotNull(nameof(coin), coin);
 			if (current.Contains(coin))
 			{
-				return current.ToList();
+				return current;
 			}
+
 			var history = current.Concat(new List<SmartCoin> { coin }).ToList(); // Fhe coin is the first elem in its history.
 
 			// If the script is the same then we have a match, no matter of the anonimity set.
-			foreach (var c in Coins)
+			foreach (var c in lookupScriptPubKey[coin.ScriptPubKey])
```

Comment:
Blockchain analysis assumes if an address is reused, then the address is reused by the same entity.

Reply:
ack.

Reply:
Sorry team but I have to insist on this point, otherwise I would feel bad with myself. Yesterday in the a Spanish bitcointalk thread someone asked me about Wasabi and online casinos and online betting sites. He made me remember that that kind of systems pay to the "address that they received from" (what is awful). So, given we cannot prevent people and services to do silly think, we should make all the possible to inform the user we useful information about the coins.

This is a possible scenario that worries me:
1) I buy one bitcoin from a miner and I label it as "Fresh coin from miner" (let say the address is bc1cleancoinxxxx)
2) I send bitcoins to an online betting site and WIN!!!
3) the betting site pays me 2 bitcoins to my address bc1cleancoinxxxx
4) I will see that something is wrong because the money arrives with a label "Fresh coin from miner". This is not very accurated but who cares! I will change that by editing the wallet.json file tomorrow, right?
5) One day I have to sell some of my 237 coins and I decide to go with the cleanest coins that I have, and there I go......

Reply:
@lontivero That is accurate. We must include "fresh coin from miner" to the history, because the "fresh coin from miner" and the "casino coin" is being associated with each other. The issue here is that Wasabi doesn't know where the "casino coin" comes from, so cannot denote that coin as "casino coin." We don't know where it is coming from, but we know that it has been associated with the "fresh coin from miner" coin, thus we show one part of the story only.

Reply:
Maybe we should let the user edit the history. Later when we will redesign coinhistory feature we should take into consideration.